ghmirror/app/__init__.py

```diff
@@ -0,0 +1,130 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2020
+# Author: Amador Pahim <apahim@redhat.com>
+
+"""
+The GitHub Mirror endpoints
+"""
+
+import logging
+import requests
+
+import flask
+
+from ghmirror.app.response import MirrorResponse
+from ghmirror.data_structures.cache import RequestsCache
+from ghmirror.data_structures.cache import StatsCache
+
+
+logging.basicConfig(level=logging.INFO,
+                    format='%(asctime)-15s %(message)s')
+
+LOG = logging.getLogger(__name__)
+GH_API = 'https://api.github.com'
+
+APP = flask.Flask(__name__)
+
+
+@APP.route('/healthz', methods=["GET"])
+def healthz():
+    """
+    Health check endpoint for Kubernetes.
+    """
+    return flask.Response('OK')
+
+
+@APP.route('/stats', methods=["GET"])
+def stats():
+    """
+    Cache statistics endpoint.
+    """
+    stats_cache = StatsCache()
+
+    return flask.jsonify(
+        {
+            'cache_hit': stats_cache.hits,
+            'cache_miss': stats_cache.misses,
+        }
+    )
+
+
+@APP.route('/', defaults={'path': ''})
+@APP.route('/<path:path>', methods=["GET", "POST", "PUT", "PATCH", "DELETE"])
+def ghmirror(path):
+    """
+    Default endpoint, matching any url without a specific endpoint.
+    """
+    url = f'{GH_API}/{path}'
+
+    if flask.request.args:
+        url += '?'
+        for key, value in flask.request.args.items():
+            url += f'{key}={value}&'
+        url = url.rstrip('&')
+
+    headers = {}
+    authorization = flask.request.headers.get('Authorization')
+    if authorization is not None:
+        headers['Authorization'] = authorization
+
+    if flask.request.method != 'GET':
+        LOG.info('[%s] BYPASS %s', flask.request.method, url)
+
+        resp = requests.request(flask.request.method,
+                                url=url,
+                                headers=headers,
+                                data=flask.request.data)
+
+        return flask.Response(resp.content,
+                              resp.status_code)
+
+    stats_cache = StatsCache()
+    requests_cache = RequestsCache()
+
+    if url in requests_cache:
+        last_modified = requests_cache[url].headers.get('Last-Modified')
+        etag = requests_cache[url].headers.get('ETag')
+        if last_modified is not None:
+            headers['If-Modified-Since'] = last_modified
+        elif etag is not None:
+            headers['If-None-Match'] = etag
+
+    resp = requests.request(flask.request.method,
+                            url=url,
+                            headers=headers)
+
+    if resp.status_code != 304:
+        LOG.info('[GET] CACHE_MISS %s', url)
+        stats_cache.miss()
+        # Caching only makes sense when at least one
+        # of those headers is present
+        if any(['Last-Modified' in resp.headers,
+                'ETag' in resp.headers]):
+            requests_cache[url] = resp
```

Comment:
perhaps at a later point we could make the key more then only the URL...
this will assist us in separating calls from different tokens (with different permissions?) for example.

Reply:
yes, I was testing the beahaviour of the api and it does generate different ETags for different users acessing the same unchanged resource, so saving the user info does make sense.